neonvm/controllers/virtualmachine_qmp_queries.go

```diff
@@ -246,57 +262,345 @@ func QmpQueryMemoryDevices(ip string, port int32) ([]QmpMemoryDevice, error) {
+				fmt.Sprintf("Failed to add dimm%d to VM %s: %s",
+					idx, r.vm.Name, err.Error()))
+			continue
+		}
+		r.recorder.Event(r.vm, "Normal", "ScaleUp",
+			fmt.Sprintf("Added dimm%d to VM %s", idx, r.vm.Name))
+		r.memBackends[idx] = true
+		r.memDevCount++
+	}
+	return attempts.DidSomething()
+}
+
+func (r *QmpMemorySetter) RemoveDevices() bool {
+	attempts := NewAttemptsCounter(r.memDevCount - r.targetCnt)
+	// Removing from the end to keep memslot1,memslot2,...
+	for idx := r.maxBackend; idx >= 1; idx-- {
+		active, ok := r.memBackends[idx]
+		if !ok || !active {
+			continue
+		}
+		// Found used backend to remove
+
+		if !attempts.Attempt() {
+			break
+		}
+
+		err := QmpDelMemoryDevice(r.mon, idx)
+		if err != nil {
+			r.errs = append(r.errs, err)
+			r.recorder.Event(r.vm, "Warning", "ScaleDown",
+				fmt.Sprintf("Failed to remove dimm%d from VM %s: %s",
+					idx, r.vm.Name, err.Error()))
+			continue
+		}
+		r.recorder.Event(r.vm, "Normal", "ScaleDown",
+			fmt.Sprintf("Removed dimm%d from VM %s", idx, r.vm.Name))
+		r.memBackends[idx] = false
+		r.memDevCount--
+	}
+
+	return attempts.DidSomething()
+}
+
+func (r *QmpMemorySetter) RemoveBackends() bool {
+	attempts := NewAttemptsCounter(len(r.memBackends) - r.targetCnt)
+
+	for idx := r.maxBackend; idx >= 1; idx-- {
+		active, ok := r.memBackends[idx]
+		if !ok || active {
+			continue
+		}
+
+		if !attempts.Attempt() {
+			break
+		}
+
+		err := QmpDelMemoryBackend(r.mon, idx)
+		if err != nil {
+			r.errs = append(r.errs, err)
+			r.recorder.Event(r.vm, "Warning", "ScaleDown",
+				fmt.Sprintf("Failed to remove memslot%d from VM %s: %s",
+					idx, r.vm.Name, err.Error()))
+			continue
+		}
+		r.recorder.Event(r.vm, "Normal", "ScaleDown",
+			fmt.Sprintf("Removed memslot%d from VM %s", idx, r.vm.Name))
+		delete(r.memBackends, idx)
+	}
+	return attempts.DidSomething()
+}
+
+func QmpSetMemorySlots(ctx context.Context, vm *vmv1.VirtualMachine, targetCnt int, recorder record.EventRecorder) (int, error) {
+	// Attempts to plug/unplug memory slots to match targetCnt.
+	// Returns the number of slots, which the function managed to plug.
+	// Ideally, it matches targetCnt, but can be less or more if there are
+	// errors.
+	//
+	// Returns -1 if failed to connect to QMP, otherwise, the return value is
+	// valid even if there are errors.
+	//
+	// In order for the hotplug to occur, we have to do two things:
+	// 1. Plug memory backend (memslot<n>) - a QEMU object, which physically allocates the
+	// memory from host
+	// 2. Plug DIMM device (dimm<n>) - a device, which exposes the memory to the
+	// host
+	//
+	// In order to unplug we need to make the same actions in the reversed order.
+	//
+	// dimm<n> is always plugged into memslot<n> with the same n.
+
+	log := log.FromContext(ctx)
+
+	setter := &QmpMemorySetter{
+		vm:        vm,
+		targetCnt: targetCnt,
+		recorder:  recorder,
+		log:       log,
+	}
+	defer setter.Disconnect()
+	setter.BuildState()
```

Comment:
The returned error is ignored here

Reply:
Right, thanks. We should have a linter for this.